tests/common/environment.rs

```diff
@@ -0,0 +1,208 @@
+#![allow(dead_code)]
+
+use super::database::{TemporaryDatabase, USER_USER_ID_PARSED};
+use crate::common::setup;
+use actix_web::{
+    dev::ServiceResponse,
+    test::{self, TestRequest},
+    App,
+};
+use chrono::Utc;
+use labrinth::{
+    database::{self, models::generate_pat_id},
+    models::pats::Scopes,
+};
+
+// A complete test environment, with a test actix app and a database.
+// Must be called in an #[actix_rt::test] context. It also simulates a
+// temporary sqlx db like #[sqlx::test] would.
+// Use .call(req) on it directly to make a test call as if test::call_service(req) were being used.
+pub struct TestEnvironment {
+    test_app: Box<dyn LocalService>,
+    pub db: TemporaryDatabase,
+}
+
+impl TestEnvironment {
+    pub async fn new() -> Self {
+        let db = TemporaryDatabase::create_with_dummy().await;
+        let labrinth_config = setup(&db).await;
+        let app = App::new().configure(|cfg| labrinth::app_config(cfg, labrinth_config.clone()));
+        let test_app = test::init_service(app).await;
+        Self {
+            test_app: Box::new(test_app),
+            db,
+        }
+    }
+    pub async fn cleanup(self) {
+        self.db.cleanup().await;
+    }
+
+    pub async fn call(&self, req: actix_http::Request) -> ServiceResponse {
+        self.test_app.call(req).await.unwrap()
+    }
+}
+
+trait LocalService {
+    fn call(
+        &self,
+        req: actix_http::Request,
+    ) -> std::pin::Pin<
+        Box<dyn std::future::Future<Output = Result<ServiceResponse, actix_web::Error>>>,
+    >;
+}
+impl<S> LocalService for S
+where
+    S: actix_web::dev::Service<
+        actix_http::Request,
+        Response = ServiceResponse,
+        Error = actix_web::Error,
+    >,
+    S::Future: 'static,
+{
+    fn call(
+        &self,
+        req: actix_http::Request,
+    ) -> std::pin::Pin<
+        Box<dyn std::future::Future<Output = Result<ServiceResponse, actix_web::Error>>>,
+    > {
+        Box::pin(self.call(req))
+    }
+}
+
+// A reusable test type that works for any scope test testing an endpoint that:
```

Comment:
I think the 2 items below this (scopetest and create pat) probably belong in a different file or `scopes.rs` and `pats.rs` respectively